src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,74 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object implementationInstance = null;
+        Class<?> clazzImplementation = findImplementation(interfaceClazz);
+        Field[] declaredFields = interfaceClazz.getDeclaredFields();
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)
+                    && clazzImplementation.isAnnotationPresent(Component.class)) {
+                Object instance = getInstance(field.getType());
+                field.setAccessible(true);
+                try {
+                    field.set(implementationInstance, instance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value: Class:"
+                            + clazzImplementation.getName() + ". Field: " + field);
+                }
+            }
+            throw new RuntimeException("Injection failed, missing @Component annotation "
+                    + "on the class" + clazzImplementation);
+        }
+        if (implementationInstance == null) {
+            implementationInstance = createInstance(clazzImplementation);
+        }
+        return implementationInstance;
+    }
+
+    private Object createInstance(Class<?> clazzImplementation) {
+        if (instances.containsKey(clazzImplementation)) {
+            return instances.get(clazzImplementation);
+        }
+        try {
+            Constructor<?> constructor = clazzImplementation.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazzImplementation, instance);
+            return instance;
+        } catch (NoSuchMethodException | InvocationTargetException
+                 | InstantiationException | IllegalAccessException e) {
```

Comment:
https://mate-academy.github.io/jv-program-common-mistakes/java-core/dependency-injection/dependency-injection#:~:text=It%20is%20better%20to%20replace%20many%20exceptions%20that%20have%20a%20common%20parent%20with%20a%20general%20parental%20exception